Describe verified boot (#72)
tillitis/dev-tillitis: hugo/content.castor/intro/_index.md

it into a programmer board.
{{% /hint %}}

## Measured and verified boot and chaining apps

The TKey creates a secret for the device application before starting
it. It does this by computing a hash function over the entire app and
mixing it together with a Unique Device Secret (UDS) embedded in the

> hardware and an optional User Supplied Secret, typically based on a

Reviewer: Introduce `USS`
```suggestion
hardware and an optional User Supplied Secret (USS), typically based on a
```

> passphrase.

Reviewer: We have since this was written, opted out of using the word `pass phrase` as describing what the user provides. The user provides a USS, User Supplied Secret.

This means that the resulting secret is based on:

1. Something the user *has*, the specific TKey device.
2. Something the user *knows*, the User Supplied Secret.
3. The *integrity* of the software, the device app they are about to
   use.

If any of these three changes, the secret is also different. If the
secret and the resulting cryptographics keys are the same as before
it's guaranteed that this is the right device, with the right software
and that the user knows the passphrase.

This is the algorithm for the secret:

```go
secret = blake2s(UDS, domain + blake2s(device_app) + USS)
```

The UDS is used as a key in a keyed version of BLAKE2s. The `domain`
here is a domain separation between if the app was using measured boot
or was chained (explained below) and with or without USS.

We call this secret the Compound Device Identity (CDI). The TKey
unconditional measured boot is inspired by, but not exactly the same
as part of [TCG
DICE](https://trustedcomputinggroup.org/work-groups/dice-architectures/).

> From this version the TKey also supports a form of verified boot in

Reviewer: ```suggestion
From codename Castor version the TKey also supports a form of verified boot in
```

order to be able to update device apps without losing key material.

In this mode the TKey uses measured boot as above to start a verifier

> app, then verifies a signature over the next app against a public key,

Reviewer: ```suggestion
app. The verifier app, verifies a signature over the next app against a public key,
```

and leaves an app digest and some data to the next app. Then it resets
the TKey.

After reset, the TKey will only accept and start and app which has the
already verified digest.

In this case, the CDI for the verified app is computed like this:

```go

> CDI = blake2s(UDS, domain + blake2s(app[n-1]'s CDI, seed)¹ + USS)

Reviewer: Maybe it's a convention I don't know about, but what does `+` mean in this context? Concatenation? What does the comma then mean?

```

> ¹ This part is actually computed by the firmware before actually doing

Reviewer: actually, actually.

the reset in order not to leak the CDI over the reset. Neither
firmware/app on any side of the reset knows about the other's CDI.

UDS is used as the key in a keyed BLAKE2s. So is `app[n-1]'s CDI`.

The `seed` is, if using Tillitis'
[tkey-boot-verifier](https://github.com/tillitis/tkey-boot-verifier) a
digest of the vendor public key used.

The `domain`, again, is a domain separation between if the app was
using measured boot or was chained (explained below) and with or
without USS.

> This means that the resuling secret/CDI is based on:

Reviewer: typo: resuling
